File: 백준/Gold/14395. 4연산/4연산.py
import sys
input = sys.stdin.readline
from collections import deque

n,m = map(int,input().split())
MAX_ = 10**9
visited = []
ans = []
cmds = ['*', '+', '-', '/']
cmds.sort()
def calc(num, cmd):
    if cmd=="+":
        return num+num
    elif cmd=="-":
        return num-num
    elif cmd=="*":
        return num*num
    else:
        return num/num


def bfs():
    global ans
    que = deque()
    que.append([n,[]])
    while que:
        num, process = que.popleft()
        for idx in range(4):
            if num==0 and cmds[idx]=="/":
                continue
            x = calc(num, cmds[idx])
    
            
            if x==m:
                #print(x, process+[cmds[idx]])
                ans.append(process+[cmds[idx]])
                return
   
            if x<=MAX_ and x not in visited:
                que.append([x, process+[cmds[idx]]])
                visited.append(x)

if n==m:
    print(0)
else:
    bfs()
    if not(ans):
        print(-1)
    else:
        ans.sort()
        answer = "".join(ans[0])
        print(answer)

File: 백준/Gold/14395. 4연산/4연산.js
const fs = require("fs")
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const inputs = fs.readFileSync(filePath).toString().split("\n")
const [n,m] = inputs[0].split(" ").map(Number)
// console.log(n,m)

const MAX_ = 10**9
const visited = []
const ans = []
const cmds=["*","+","-","/"]
cmds.sort()


function calc(num, cmd){
  if (cmd==="+"){
    return num+num
  }
  else if (cmd==="-"){
    return num-num
  }
  else if (cmd==="*"){
    return num * num
  }
  else if(cmd=="/"){
    return num/num
  }
}


function bfs(){
  const que = [[n, []]]
  while(que.length>0){
    const [num,process] = que.shift()
    
    for(let cmd of cmds){
      //console.log(num, process, cmd, cmds);
      if (num===0 && cmd==="/"){
        continue
      }
      x = calc(num, cmd)

      // 목표 수 도달한 경우
      if (x===m){
        // console.log([...process, cmd]);
        ans.push([...process, cmd])
        return
      }
      if((x<=MAX_)&&(!visited.includes(x))){
        // console.log([...process, cmd]);
        que.push([x, [...process, cmd]])
        visited.push(x)
      }
    }
  }
}
bfs()
if (n===m){
  console.log(0)
}
else{
  if(ans.length===0){
    console.log(-1)
  }
  else{
    ans.sort()
    console.log(ans[0].join(""));
  }
}
